xcdat/dataset.py

```diff
@@ -583,27 +453,167 @@ def _postprocess_dataset(
+    ``xr.coding.variables.lazy_elemwise_func`` can be leveraged to lazily decode
+    time.
 
-    return dataset
+    This function is based on ``xarray.coding.times.CFDatetimeCoder.decode``.
+
+    Parameters
+    ----------
+    coords : xr.DataArray
+        A DataArray of numerically encoded time.
+
+    Returns
+    -------
+    xr.Variable
+        A Variable of time decoded as ``cftime`` objects.
+    """
+    variable = as_variable(da)
+    dims, data, attrs, encoding = unpack_for_decoding(variable)
+
+    units = pop_to(attrs, encoding, "units")
+    calendar = pop_to(attrs, encoding, "calendar")
+
+    transform = partial(_get_cftime_coords, units=units, calendar=calendar)
+    data = lazy_elemwise_func(data, transform, np.dtype("object"))
+
+    return xr.Variable(dims, data, attrs, encoding)
+
+
+def _get_cftime_coords(offsets: np.ndarray, units: str, calendar: str) -> np.ndarray:
+    """Get an array of `cftime` coordinates starting from a reference date.
+
+    This function supports multidimensional arrays such as bounds by flattening
+    them first, decoding, then reshaping back to their original shapes.
+
+    Parameters
+    ----------
+    offsets : np.ndarray
+        An array of numerically encoded time offsets from the reference date.
+    calendar : str
+        The CF calendar type supported by ``cftime`` . This includes "noleap",
+        "360_day", "365_day", "366_day", "gregorian", "proleptic_gregorian",
+        "julian", "all_leap", and "standard".
+    units : str
+        The time units.
+
+    Returns
+    -------
+    np.ndarray
+        An array of `cftime` coordinates.
+    """
+    units_type, ref_date = units.split(" since ")
+
+    if units_type in NON_CF_TIME_UNITS:
+        offsets = np.asarray(offsets)
+        flat_offsets = offsets.ravel()
+
+        # Convert offsets to `np.float64` to avoid "TypeError: unsupported type
+        # for timedelta days component: numpy.int64".
+        flat_offsets = flat_offsets.astype("float")
+
+        # We don't need to do calendar arithmetic here because the units and
+        # offsets are in "months" or "years", which means leap days should not
+        # be factored.
+        ref_datetime: datetime = parser.parse(ref_date, default=datetime(2000, 1, 1))
+        times = np.array(
+            [
+                ref_datetime + rd.relativedelta(**{units_type: offset})
+                for offset in flat_offsets
+            ],
+            dtype="object",
+        )
+        # Convert the array of `datetime` objects into `cftime` objects based on
+        # the calendar type.
+        date_type = get_date_type(calendar)
+        coords = convert_times(times, date_type=date_type)
+
+        # Reshape back to the original shape.
+        coords = coords.reshape(offsets.shape)
+    else:
+        xrunits = units + " since " + ref_date
+        coords = xr.coding.times.decode_cf_datetime(
+            offsets, xrunits, calendar=calendar, use_cftime=True
+        )
```

Comment:
```suggestion
        coords = xr.coding.times.decode_cf_datetime(
            offsets, units, calendar=calendar, use_cftime=True
        )
```

We can reuse the `units` function arg because it is already in the original format, instead of recreating it as `xrunits`

I'm also not sure why `decode_cf_datetime()` didn't break because the `xrunits` formula is incorrect.
The formula should have been `xrunits = units_type + " since " + ref_date`. I'll take a look closer look.

Reply:
This section of code works fine despite being incorrect because `decode_cf_datetime()` calls `cftime.num2date()`, which has a function called `_datesplit()`.

If we incorrectly set `xrunits = "days since 2001-0-01 since 2001-01`:
1. `_date_split()` splits the units string on ` "since", resulting in `['days', 'since', '2000-01-01 since 2000-01-01']`.
https://github.com/Unidata/cftime/blob/5d5271b664bc55d33f91bfd6b96923f2f14d2c22/src/cftime/_cftime.pyx#L71-L83
2. Only the `days` sub-string is used, which explains why our code works fine.
https://github.com/Unidata/cftime/blob/master/src/cftime/_cftime.pyx#L588

Although the code works, I still updated it to my suggestion since that is the intended behavior.